### ISSUE TITLE
Unbounded memory usage

### ISSUE THREAD

**Issue by mharbison72** [google/pytype#860]
I'm adding typehints to Mercurial using WSL Ubuntu 18.04 with python 3.8.0, and the 2021.03.10 release seems to get stuck about 2/3 of the way through ([245/339] on my system).  Watching in task manager, the python38 process climbed up to 4GB and was using ~20% cpu consistently before I killed it.

To reproduce this, expand the attached tarball, and run the following from within the directory:

```
    make local >/dev/null
    cd tests && time ./run-tests.py --local --allow-slow-tests test-check-pytype.t --debug
```

The file that it appears to hang on is random, but upon killing the runaway process, the file it was actually stuck on appears pretty consistent:

```
FAILED: /mnt/c/Users/Matt/pytype-hang/hg-pytype-loop/.pytype/pyi/mercurial/upgrade_utils/actions.pyi-1
/usr/bin/python3.8 -m pytype.single --imports_info /mnt/c/Users/Matt/pytype-hang/hg-pytype-loop/.pytype/imports/mercurial.upgrade_utils.actions.imports-1 --module-name mercurial.upgrade_utils.actions -V 3.6 -o /mnt/c/Users/Matt/pytype-hang/hg-pytype-loop/.pytype/pyi/mercurial/upgrade_utils/actions.pyi-1 --no-report-errors --nofail --quick /mnt/c/Users/Matt/pytype-hang/hg-pytype-loop/mercurial/upgrade_utils/actions.py
Killed
```

It's entirely possible that there are errors in the type hints, but this is a recent regression, as I've been using 2020.12.02 without issue.  It looks like the latest release that runs to completion is 2021.02.09.

[hg-pytype-loop.gz](https://github.com/google/pytype/files/6146154/hg-pytype-loop.gz)

**Comment by mharbison72** [comment]
Not sure if this will help, but I reproduced this on CentOS 7 with py3.6.8, and got the following from py-spy on the spinning process:

```
$ sudo ~/.local/bin/py-spy top --pid 15652

Total Samples 3058
GIL: 100.00%, Active: 100.00%, Threads: 1

  %Own   %Total  OwnTime  TotalTime  Function (filename:line)
100.00% 100.00%   30.58s    30.58s   _get_values (pytype/output.py:87)
  0.00% 100.00%   0.000s    30.58s   _run_pytype (pytype/single.py:93)
  0.00% 100.00%   0.000s    30.58s   wrapper (pytype/io.py:63)
  0.00% 100.00%   0.000s    30.58s   compute_types (pytype/analyze.py:598)
  0.00% 100.00%   0.000s    30.58s   value_to_pytd_type (pytype/output.py:293)
  0.00% 100.00%   0.000s    30.58s   _function_call_combination_to_signature (pytype/output.py:454)
  0.00% 100.00%   0.000s    30.58s   to_pytd_def (pytype/abstract.py:263)
  0.00% 100.00%   0.000s    30.58s   _call (pytype/io.py:81)
  0.00% 100.00%   0.000s    30.58s   _run_code (runpy.py:85)
  0.00% 100.00%   0.000s    30.58s   value_instance_to_pytd_type (pytype/output.py:203)
  0.00% 100.00%   0.000s    30.58s   infer_types (pytype/analyze.py:692)
  0.00% 100.00%   0.000s    30.58s   _value_to_parameter_types (pytype/output.py:125)
  0.00% 100.00%   0.000s    30.58s   <module> (pytype/single.py:97)
  0.00% 100.00%   0.000s    30.58s   <genexpr> (pytype/output.py:492)
  0.00% 100.00%   0.000s    30.58s   _function_to_def (pytype/output.py:492)
  0.00% 100.00%   0.000s    30.58s   generate_pyi (pytype/io.py:109)
  0.00% 100.00%   0.000s    30.58s   to_type (pytype/abstract.py:259)
  0.00% 100.00%   0.000s    30.58s   _function_call_to_return_type (pytype/output.py:432)
  0.00% 100.00%   0.000s    30.58s   _run_module_as_main (runpy.py:193)
  0.00% 100.00%   0.000s    30.58s   process_one_file (pytype/io.py:202)
  0.00% 100.00%   0.000s    30.58s   pytd_for_types (pytype/analyze.py:506)
  0.00% 100.00%   0.000s    30.58s   check_or_generate_pyi (pytype/io.py:151)
  0.00% 100.00%   0.000s    30.58s   main (pytype/single.py:82)
  0.00% 100.00%   0.000s    30.58s   value_to_pytd_def (pytype/output.py:389)

Press Control-C to quit, or ? for help.

$ sudo ~/.local/bin/py-spy dump --pid 15652
Process 15652: /bin/python3 -m pytype.single --imports_info /home/jenkins/hg-pytype-loop/.pytype/imports/mercurial.upgrade_utils.actions.imports-1 --module-name mercurial.upgrade_utils.actions -V 3.6 -o /home/jenkins/hg-pytype-loop/.pytype/pyi/mercurial/upgrade_utils/actions.pyi-1 --no-report-errors --nofail --quick /home/jenkins/hg-pytype-loop/mercurial/upgrade_utils/actions.py
Python v3.6.8 (/usr/bin/python3.6)

Thread 15652 (active+gil): "MainThread"
    _get_values (pytype/output.py:87)
    _value_to_parameter_types (pytype/output.py:125)
    value_instance_to_pytd_type (pytype/output.py:203)
    value_to_pytd_type (pytype/output.py:293)
    to_type (pytype/abstract.py:259)
    _function_call_to_return_type (pytype/output.py:432)
    _function_call_combination_to_signature (pytype/output.py:454)
    <genexpr> (pytype/output.py:492)
    _function_to_def (pytype/output.py:492)
    value_to_pytd_def (pytype/output.py:389)
    to_pytd_def (pytype/abstract.py:263)
    pytd_for_types (pytype/analyze.py:506)
    compute_types (pytype/analyze.py:598)
    infer_types (pytype/analyze.py:692)
    _call (pytype/io.py:81)
    wrapper (pytype/io.py:63)
    generate_pyi (pytype/io.py:109)
    check_or_generate_pyi (pytype/io.py:151)
    wrapper (pytype/io.py:63)
    process_one_file (pytype/io.py:202)
    wrapper (pytype/io.py:63)
    _run_pytype (pytype/single.py:93)
    main (pytype/single.py:82)
    <module> (pytype/single.py:97)
    _run_code (runpy.py:85)
    _run_module_as_main (runpy.py:193)
```

**Comment by rchen152** [comment]
Sorry for the slow response! If pytype-2021.02.09 is the last release that runs to completion, then the breaking change is likely https://github.com/google/pytype/commit/33bfb9b634d3563a0c81fa3f28665d46e51eb8f1. That commit fixed a bug in pytype's handling of cycles in the control flow graph it constructs, with the side effect that it occasionally has to do a lot more work to answer reachability queries that were previously (erroneously) satisfied by a cycle.

I'll poke at this a bit more to see if I can find a particular reason that pytype is hanging, but if I'm right and it's a CFG issue, there may not be much I can do.

Glancing at [mercurial/upgrade_utils/actions.py](https://www.mercurial-scm.org/repo/hg/file/tip/mercurial/upgrade_utils/actions.py), it doesn't strike me as particularly long or complex, but sometimes splitting up a file can help. The nuclear option is to slap a `# pytype: skip-file` at the top of the file to make pytype skip analysis and treat everything in the file as having type `Any`.

**Comment by mharbison72** [comment]
Thanks for the reply.  I can confirm that f2fea9c works and 33bfb9b fails.

Is there any syntax for skipping sections of a file instead of the whole thing?  Similar to `# pytype: disable=...` followed by `# pytype: enable=...` several lines later for specific errors.  That might help narrow down which part of the file it is having a problem with.  I let this run for 50 minutes and it ran up to 21GB of memory before I had to kill it due to running low on swap space.  So I'm not sure that it would ever complete.

**Comment by rchen152** [comment]
Ah, unfortunately we don't have anything like that. `skip-file` is the only directive that actually disables analysis; the others just turn off reporting of a particular error.

**Comment by rchen152** [comment]
Ok, I narrowed the issue down somewhat, and I think I've found a workaround.

**tl;dr** The workaround is to add a type annotation to the line that initializes `allformatvariant`:
```
allformatvariant = []  # type: list[type['formatvariant']]
```
(line 110 in `mercurial/upgrade_utils/actions.py` in the tarball)

**Longer explanation** I reduced `actions.py` down slightly (see [actions.py.zip](https://github.com/google/pytype/files/6300745/actions.py.zip) ) to discover that the following command:
```
pytype-single -o - --quick actions.py --timeout 15 -v4
```
hangs until killed by `--timeout 15`. But if I delete the `find_format_downgrades` method (lines 444-463), the pytype run completes in about 2s. `find_format_downgrades` loops over `allformatvariant`, a module-level list of `formatvariant` subclasses. A known issue with pytype (that we're working on this quarter, actually!) is that it takes a long time to analyze large concrete data structures. You can work around this issue by type-annotating the data structure, which causes pytype to create an abstract instance of it rather than tracking its literal value.

I'm going to go ahead and close this issue, since its root cause in the end appears to be something that's already on our radar and actively being worked on. But let me know if anything is unclear or if the suggested workaround doesn't work.

**Comment by mharbison72** [comment]
That did the trick, thanks!

PyCharm complained that "`type` doesn't define `__getitem__`, so `[]` can't be used".  But I see in the generated pyi file that pytype changed that to `Type` and PyCharm is happy with that.  So I assume that the ability to use `type` there is a very recent change?

**Comment by rchen152** [comment]
Ah yes, my bad. For older Python 3 versions, it's better to use `typing.List` and `typing.Type` for now.
